clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -974,49 +984,36 @@ mlir::LogicalResult CIRGenFunction::buildSwitchStmt(const SwitchStmt &S) {
     // TODO: PGO and likelihood (e.g. PGO.haveRegionCounts())
     // TODO: if the switch has a condition wrapped by __builtin_unpredictable?
 
-    // FIXME: track switch to handle nested stmts.
     swop = builder.create<SwitchOp>(
         getLoc(S.getBeginLoc()), condV,
         /*switchBuilder=*/
         [&](mlir::OpBuilder &b, mlir::Location loc, mlir::OperationState &os) {
-          auto *cs = dyn_cast<CompoundStmt>(S.getBody());
-          assert(cs && "expected compound stmt");
-          SmallVector<mlir::Attribute, 4> caseAttrs;
-
           currLexScope->setAsSwitch();
-          mlir::Block *lastCaseBlock = nullptr;
-          for (auto *c : cs->body()) {
-            bool caseLike = isa<CaseStmt, DefaultStmt>(c);
-            if (!caseLike) {
-              // This means it's a random stmt following up a case, just
-              // emit it as part of previous known case.
-              assert(lastCaseBlock && "expects pre-existing case block");
-              mlir::OpBuilder::InsertionGuard guardCase(builder);
-              builder.setInsertionPointToEnd(lastCaseBlock);
-              res = buildStmt(c, /*useCurrentScope=*/!isa<CompoundStmt>(c));
-              lastCaseBlock = builder.getBlock();
-              if (res.failed())
-                break;
-              continue;
-            }
+          res = buildSwitchBody(S.getBody());
 
-            auto *caseStmt = dyn_cast<CaseStmt>(c);
+          os.addRegions(currLexScope->getSwitchRegions());
 
-            if (caseStmt)
-              res = buildCaseStmt(*caseStmt, condV.getType(), caseAttrs, os);
-            else {
-              auto *defaultStmt = dyn_cast<DefaultStmt>(c);
-              assert(defaultStmt && "expected default stmt");
-              res = buildDefaultStmt(*defaultStmt, condV.getType(), caseAttrs,
-                                     os);
+          SmallVector<mlir::Attribute, 4> caseAttrs;
```

Comment:
This block of logic here can be abstracted into its own helper function: all it does it building `CaseOpKindAttr`.

Reply:
Added `switchCondType` in `LexScope` and moved these logic into helper function, thanks.

Reply:
Passing params to functions is totally fine, things shouldn't be added to LexScope unless strictly needed. And if it's needed to fix a bug or support nested switches, should be added when time for that PR comes. Please don't add `switchCondType` into `LexScope`.

Reply:
Sure, reverted the related changes, leave it to the future pr.